cpp/pmft/PMFTXY2D.cc

```diff
@@ -41,177 +41,91 @@ PMFTXY2D::PMFTXY2D(float x_max, float y_max, unsigned int n_x, unsigned int n_y)
-        [=] (const blocked_range<size_t>& r)
-            {
-            Index2D b_i = Index2D(m_n_x, m_n_y);
-            for (size_t i = r.begin(); i != r.end(); i++)
-                {
-                for (size_t j = 0; j < m_n_y; j++)
-                    {
-                    for (tbb::enumerable_thread_specific<unsigned int *>::const_iterator local_bins = m_local_bin_counts.begin();
-                         local_bins != m_local_bin_counts.end(); ++local_bins)
-                        {
-                        m_bin_counts.get()[b_i((int)i, (int)j)] += (*local_bins)[b_i((int)i, (int)j)];
-                        }
-                    }
-                }
-            });
-    float inv_num_dens = this->m_box.getVolume() / (float)this->m_n_p;
-    float inv_jacobian = (float) 1.0 / m_jacobian;
-    float norm_factor = (float) 1.0 / ((float) this->m_frame_counter * (float) this->m_n_ref);
-    // normalize pcf_array
-    parallel_for(blocked_range<size_t>(0,m_n_x*m_n_y),
-        [=] (const blocked_range<size_t>& r)
-            {
-            for (size_t i = r.begin(); i != r.end(); i++)
-                {
-                m_pcf_array.get()[i] = (float)m_bin_counts.get()[i] * norm_factor * inv_jacobian * inv_num_dens;
-                }
-            });
-    }
+{
+    float jocob_factor = (float) 1.0 / m_jacobian;
+    reduce2D(m_n_x, m_n_y, [jocob_factor](size_t i) { return jocob_factor; });
+}
 
 //! \internal
 /*! \brief Function to reset the pcf array if needed e.g. calculating between new particle types
-*/
+ */
 
 void PMFTXY2D::reset()
-    {
-    for (tbb::enumerable_thread_specific<unsigned int *>::iterator i = m_local_bin_counts.begin(); i != m_local_bin_counts.end(); ++i)
-        {
-        memset((void*) (*i), 0, sizeof(unsigned int)*m_n_x*m_n_y);
-        }
-    this->m_frame_counter = 0;
-    this->m_reduce = true;
-    }
+{
+    resetGeneral(m_n_x * m_n_y);
+}
 
 //! \internal
 /*! \brief Helper functionto direct the calculation to the correct helper class
-*/
+ */
+
+void PMFTXY2D::accumulate(box::Box& box, const locality::NeighborList* nlist, vec3<float>* ref_points,
+                          float* ref_orientations, unsigned int n_ref, vec3<float>* points,
+                          float* orientations, unsigned int n_p)
+{
+    assert(ref_points);
+    assert(points);
+    assert(n_ref > 0);
+    assert(n_p > 0);
+
+    // precalc some values for faster computation within the loop
+    float dx_inv = 1.0f / m_dx;
+    float dy_inv = 1.0f / m_dy;
+
+    Index2D b_i = Index2D(m_n_x, m_n_y);
+
+    accumulateGeneral(box, n_ref, nlist, n_p, m_n_x * m_n_y, [=](size_t i, size_t j) {
+        vec3<float> ref = ref_points[i];
+        vec3<float> delta = this->m_box.wrap(points[j] - ref);
+        float rsq = dot(delta, delta);
+
+        // check that the particle is not checking itself
+        // 1e-6 is an arbitrary value that could be set differently if needed
+        if (rsq < 1e-6)
```

Comment:
Oh gosh. This needs to be handled differently with `exclude_ii`-type logic.

Reply:
Is it guaranteed that if the indices of two points are same, then the two points are same points?

Reply:
We are fixing this in #328.